api/scpca_portal/models/library.py

```diff
@@ -175,21 +189,58 @@ def get_metadata(self) -> Dict:
-            for sample in self.samples
-        ]
+        return library_metadata
 
-    def get_filtered_data_file_paths(self, download_config: Dict) -> List[Path]:
-        file_extension_to_be_excluded = (
-            common.ANNDATA_EXT
-            if download_config["format"] == Library.FileFormats.SINGLE_CELL_EXPERIMENT
-            else common.SCE_EXT
-        )
+    def get_combined_library_metadata(self) -> List[Dict]:
+        combined_metadatas = []
+        for sample in self.samples.all():
+            metadata = self.project.get_metadata() | sample.get_metadata() | self.get_metadata()
+            # Estimate attributes per modality:
+            #   Single Cell: "sample_cell_count_estimate"
+            #   Single Cell Multiplexed: "sample_cell_estimates"
+            #   Spatial: None
+            if self.modality == Library.Modalities.SPATIAL or self.is_multiplexed:
+                del metadata["sample_cell_count_estimate"]
+            if not self.is_multiplexed:
+                del metadata["sample_cell_estimates"]
+
+            combined_metadatas.append(metadata)
+
+        return combined_metadatas
+
+    def get_download_config_file_paths(self, download_config: Dict) -> List[Path]:
+        """
+        Return all of a library's file paths that are suitable for the passed download config.
+        """
+        omit_suffixes = set()
+        # If metadata_only download, then leave omit suffixes as empty set, else populate
+        if not download_config.get("metadata_only", False):
+            omit_suffixes = set(common.FORMAT_EXTENSIONS.values())
+            # If merged download, keep omit suffixes set full as no library data files are included
+            # If not, remove download config format from omit suffixes set
+            if not download_config.get("includes_merged", False):
+                omit_suffixes.remove(common.FORMAT_EXTENSIONS.get(download_config["format"]))
 
         return [
             file_path
             for file_path in [Path(fp) for fp in self.data_file_paths]
-            if file_path.suffix != file_extension_to_be_excluded
+            if file_path.suffix not in omit_suffixes
         ]
+
+    @staticmethod
+    def get_local_file_path(file_path: Path):
+        return common.INPUT_DATA_PATH / file_path
+
+    @staticmethod
+    def get_zip_file_path(file_path: Path, download_config: Dict) -> Path:
+        path_parts = [Path(path) for path in file_path.parts]
+
+        # Project output paths are relative to project directory
+        if "metadata_only" in download_config:
+            output_path = file_path.relative_to(path_parts[0])
+        # Sample output paths are relative to project and sample directories
+        else:
+            output_path = file_path.relative_to(path_parts[0] / path_parts[1])
+
+        # Comma separated lists of multiplexed samples should become underscore separated
+        return Path(str(output_path).replace(",", "_"))
```

Comment:
Wanted to ask your opinion on this @davidsmejia.

At some point you mentioned transitioning towards one `create_computed_files` method from the two `Project::get_computed_files` and `Sample::get_computed_files`, and one `get_file` method from the two `ComputedFile::get_project_file` and `Computed_File::get_sample_file`.

Obviously line 239 of checking for "metadata_only" is non-obvious that this field only belongs to `Project` download configs. In line with the above point, would it make sense to add to each `download_config` a `type` field, which would have a value of either `"Project"` or `"Sample"`? If so, then that could be used as the conditional identifier in line 239 here as well.

Reply:
I think the more correct way to check this is if `download_config` is in `common.GENERATED_PROJECT_DOWNLOAD_CONFIGURATIONS`. That would help clear up the ambiguity.

The other thing is this download config will change when we move to ad hoc but it is too early to tell what that should look like now. Most likely we will be passing in the project/sample/queryset and getting the libraries from that. But again not a problem for today.

Reply:
Yes! A clean solution which is relevant today. Very nice. We'll solve the bigger problems soon enough!